packages/ckeditor5-link/src/linkcommand.js

```diff
@@ -187,18 +187,49 @@ export default class LinkCommand extends Command {
 				}
 			} else {
 				// If selection has non-collapsed ranges, we change attribute on nodes inside those ranges
-				// omitting nodes where `linkHref` attribute is disallowed.
+				// omitting nodes where the `linkHref` attribute is disallowed.
 				const ranges = model.schema.getValidRanges( selection.getRanges(), 'linkHref' );
 
+				// But for the first, check whether the `linkHref` attribute is allowed on selected blocks (e.g. the "image" element).
+				const allowedRanges = [];
+
+				for ( const element of selection.getSelectedBlocks() ) {
+					if ( model.schema.checkAttribute( element, 'linkHref' ) ) {
+						allowedRanges.push( writer.createRangeOn( element ) );
+					}
+				}
+
+				// Ranges that accept the `linkHref` attribute. Since we will iterate over `allowedRanges`, let's clone it.
+				const rangesToUpdate = allowedRanges.slice();
+
+				// For all ranges we want to check whether given range is inside an element that accepts the `linkHref` attribute.
+				// If so, we don't want to propagate applying the attribute to its children.
 				for ( const range of ranges ) {
-					writer.setAttribute( 'linkHref', href, range );
+					let isRangeToUpdate = true;
+
+					for ( const allowedRange of allowedRanges ) {
```

Comment:
Could you extract this as a method, like `isRangeToUpdate( range, allowedRanges )`. It will simplify that loop a bit.

Another thing is that it could even make 202-221 shorter (I didn't test the below though):

```javascript
const rangesToUpdate = [ 
    ...allowedRanges,
    ...ranges.filter( range => isRangeToUpdate( range, allowedRanges ) )
];
```

Reply:
The shorter syntax you proposed looks nice. Unfortunately, it does not work so we need to go with the `for of` loop. After extracting its internals to a separate method it looks even clear.

Reply:
Yeah, I ddin't test that :D Anyway now looks nicer :)

Reply:
Yeah, I ddin't test that :D Anyway now looks nicer :)